### PR TITLE
Adding an iterative version of prefix tree

### PR DESCRIPTION
**Pull request by cpurmessur**
I added an iterative version of prefix tree to improve performance and reduce ram usage as compared to when the recursive version is used. I also added some tests to confirm that both versions output the same directed prefix tree graph.

Note: I initially made changes to cliques.py by mistake but I undid those changes in a commit and now only 2 files have been changed (trees.py and test_trees.py)

Thanks, @dschult for your help with this

### CODE REVIEW

File: networkx/generators/tests/test_trees.py [hunk 1]
```diff
@@ -67,6 +67,86 @@ def source_label(v):
         assert source_label(ten) == "n"
         assert source_label(NIL) == "NIL"
 
+    def test_basic_recursive(self):
```
Comment (rossbar) [inline networkx/generators/tests/test_trees.py:70]: AFAICT this test is identical to the previous one - in that case, you could instead parametrize the test on the prefix tree function - something like:

```python
@pytest.mark.parametrize("prefix_tree_fn", (nx.prefix_tree, nx.prefix_tree_recursive))
def test_prefix_tree(self, prefix_tree_fn):
    ...
    T = prefix_tree_fn(strings)
    ...
```

This has the advantage of:
 1. Making the code significantly shorter, and more importantly
 2. Reducing duplication so that any future modification to the test doesn't have to be done in two separate places.

File: networkx/generators/tests/test_trees.py [hunk 2]
```diff
@@ -67,6 +67,86 @@ def source_label(v):
...
+        strings = ["ab", "abs", "ad"]
+        T = nx.prefix_tree(strings)
+        R = nx.prefix_tree_recursive(strings)
+        nx.testing.assert_graphs_equal(T, R)
+
+    def test_iterative_and_recursive_equivalence3(self):
+        strings = ["distant", "disparaging", "distant", "diamond", "ruby"]
+        T = nx.prefix_tree(strings)
+        R = nx.prefix_tree_recursive(strings)
+        nx.testing.assert_graphs_equal(T, R)
```
Comment (rossbar) [inline networkx/generators/tests/test_trees.py:148]: Similar argument here, since the only thing changing for these tests is the inputs:

```suggestion
    @pytest.mark.parametrize(
        "strings",
        (
            ["ab", "abs", "ad"],
            ["distant", "disparaging", "distant", "diamond", "ruby"],
        )
    )
    def test_implementations_consistent(self, strings):
        """Ensure results are consistent between prefix_tree implementations."""
        nx.testing.assert_graphs_equal(
            nx.prefix_tree(strings),
            nx.prefix_tree_recursive(strings),
        )
```

This makes it easier to add new test-cases (if desired). Also I left off the first example since it seems to be redundant with the inputs in the previous test(s).

**Review by rossbar — COMMENTED** [review]
I haven't taken a detailed look at the new implementation yet, but a couple things jumped out about how the tests could be organized. No blockers or anything, just a couple suggestions for reducing a bit of redundancy.

I aim to get around to the new implementation soon (and maybe using it to test some benchmarking frameworks - it seems like a nice candidate for that!)

**Comment by cpurmessur** [comment]
I have made some changes to the previous code.

1. Testing
@rossbar  thanks for the feedback on testing. I have parameterized the tests as you suggested. It is much cleaner and less redundant now. Note that I did include the first example in the second test as well because it is possible for the first test to give a positive result with a particular example while the second test gives a negative result with the same example. Essentially, the first test ensures that the tree has the correct branches, and the second test ensures that the naming of the nodes of the tree in the recursive version is the same as the one in the iterative version(s).

2. Implementation of Algorithm
I have added another version of the prefix tree implementation (prefix_tree_alt). The first version is a little easier to read but it does have some repetition and redundancy. The second version is a little bit more difficult to follow but there is no repetition. The second version is 3-4 lines shorter. Let me know which version you think is best and I can erase the other one and push in another commit. Thanks!

File: networkx/generators/trees.py [hunk 3]
```diff
@@ -29,6 +29,277 @@ def prefix_tree(paths):
...
+    tree.add_node(NIL, source="NIL")
+    stack = []
+    children = defaultdict(list)
+    # Populate dictionary with the key(s) as the child/children of the root and
+    # the value(s) as the remaining paths of the corresponding child/children
+    for path in paths:
+        # If path is empty, we add an edge to the NIL node.
+        if not path:
+            tree.add_edge(root, NIL)
+            continue
```
Comment (rossbar) [inline networkx/generators/trees.py:114]: Since there's a special case here in the initialization for empty containers, it might be good to add a test case to the parametrized tests with an empty string in it, just to make sure that the behavior is consistent with the recursive formulation. In other words, adding `["ab", "abs", "ad", ""]` to the parametrization list for for `test_implementations_consistent`.

**Review by rossbar — COMMENTED** [review]
Wow, this is a pretty nifty implementation with some slick ideas - very nice! It took me a while to grok and I may not be 100% there but hopefully my suggestions aren't way off base.

FWIW I prefer the implementation in `prefix_tree` to the `prefix_tree_alt`. Iterating over a stack that is modified by the loop was easier for me to follow than tracing the decision branches based on the `addpaths` flag. Also, if my comments re: the chunk of code for constructing `children` isn't way off, it may be possible to make the `prefix_tree` implementation even simpler and improve code reuse, which would remove some of the motivation for the alternative implementation.

**Comment by dschult** [comment]
This looks pretty good.  You can click the "Resolve Conversation" button for any comments that your changes have addressed. Leave it open if it wasn't fully addressed and add a "reply" comment.
Thanks for this!

**Comment by cpurmessur** [comment]
I think I have made all the changes. Thanks, @dschult and @rossbar  for your help with this! Let me know if you have any other suggestions

**Review by dschult — APPROVED** [review]
This looks good with the suggested function-in-a-function idea @rossbar suggested. 
Thanks for this @cpurmessur !

**Review by rossbar — APPROVED** [review]
Very nice! I'm going to add a note to the `Improvements` section of the release notes then will merge once CI clears. Thanks @cpurmessur !
